src/utility/png.cpp

```diff
@@ -0,0 +1,190 @@
+/**
+ * Copyright (c) 2011-2015 libbitcoin developers (see AUTHORS)
+ *
+ * This file is part of libbitcoin.
+ *
+ * libbitcoin is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License with
+ * additional permissions to the one published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version. For more information see LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <bitcoin/bitcoin/utility/png.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <iostream>
+#include <boost/iostreams/stream.hpp>
+#include <bitcoin/bitcoin/formats/base16.hpp>
+#include <bitcoin/bitcoin/utility/container_sink.hpp>
+#include <bitcoin/bitcoin/utility/container_source.hpp>
+#include <bitcoin/bitcoin/utility/istream_reader.hpp>
+#include <bitcoin/bitcoin/utility/ostream_writer.hpp>
+
+namespace libbitcoin {
+
+#ifdef WITH_PNG
+
+bool write_png(const data_chunk& data, const uint32_t size,
+    std::ostream& out)
+{
+    data_source istream(data);
+    return write_png(istream, size, out);
+}
+
+bool write_png(const data_chunk& data, const uint32_t size,
+    const uint32_t dpi, const uint32_t margin, const uint32_t inches_per_meter,
+    const png::color foreground, const png::color background,
+    std::ostream& out)
+{
+    data_source istream(data);
+    return write_png(istream, size, dpi, margin, inches_per_meter,
+        foreground, background, out);
+}
+
+bool write_png(std::istream& in, const uint32_t size,
+    std::ostream& out)
+{
+    return write_png(in, size, png::dpi, png::margin,
+        png::inches_per_meter, png::foreground, png::background, out);
+}
+
+extern "C" void internal_png_sink_write(png_structp png_ptr,
+    png_bytep data, png_size_t length)
+{
+    ostream_writer& sink = *reinterpret_cast<ostream_writer*>(
+        png_get_io_ptr(png_ptr));
+    sink.write_data(reinterpret_cast<const uint8_t*>(data), size_t(length));
+}
+
+extern "C" void user_error_fn(png_structp png_ptr, png_const_charp error_msg)
+{
+    throw std::runtime_error(error_msg);
+}
+
+bool write_png(std::istream& in, const uint32_t size,
+    const uint32_t dpi, const uint32_t margin, const uint32_t inches_per_meter,
+    const png::color foreground, const png::color background,
+    std::ostream& out)
+{
+    uint32_t version, width;
+    int32_t x, y, xx, yy, bit;
+    const auto bits_per_byte = 8;
```

Comment:
Use `static constexpr` here. Note that this will be typed as `int` so better to be explicit in most integer declarations (i.e. `int32_t`) if that is the intent.